### PR TITLE
Allow to use operator tokens as functions

### PR DESCRIPTION
**Pull request by ckganesan**
This PR enables the usage of operator tokens as functions. #572

### CODE REVIEW

**Comment by antonmedv** [comment]
Lets just allow to use "endsWith"
"startsWith"
"contains"
"matches" words as method names?

**Comment by ckganesan** [comment]
> Lets just allow to use "endsWith" "startsWith" "contains" "matches" words as method names?

I have made the necessary updates based on your feedback.

**Comment by Q16G** [comment]
This function is very important for my code, could you please add this function to the latest version with a tag, so that I can add it to the code

**Comment by antonmedv** [comment]
Let's try to implement this feature without adding additional options. I will try to do it.

**Comment by ckganesan** [comment]
> Let's try to implement this feature without adding additional options. I will try to do it.

I have option functionality from configuration as per your feedback

File: parser/parser_test.go [hunk 1]
```diff
@@ -342,12 +349,24 @@ world`},
...
+				Right: &StringNode{Value: "foo"}},
+		},
 		{
 			`foo endsWith "foo"`,
 			&BinaryNode{Operator: "endsWith",
 				Left:  &IdentifierNode{Value: "foo"},
 				Right: &StringNode{Value: "foo"}},
 		},
+		{
+			`endsWith(foo,  "foo")`,
```
Comment (antonmedv) [inline parser/parser_test.go:365]: Let's also add test for postfix/prefix and other combinations as:
```
endsWith() endsWith endsWith()
endsWith().Field
endsWith()[0]
-+endsWith()
endsWith()+endsWith()
```
To see how parser whould behaive in complex scenarios?

File: parser/parser_test.go [hunk 2]
```diff
@@ -342,12 +349,24 @@ world`},
...
+			&BinaryNode{Operator: "endsWith",
+				Left:  &IdentifierNode{Value: "foo"},
+				Right: &StringNode{Value: "foo"}},
+		},
```
Comment (antonmedv) [inline parser/parser_test.go:369]: And lets also add a test withing expr_test to see if it is possible to add own function there?

File: parser/parser_test.go (hunk 1 above)
Comment (ckganesan) [reply parser/parser_test.go]: added more tests cases as per your feedback

**Review by antonmedv — APPROVED** [review]

**Comment by antonmedv** [comment]
Awesome work!

**Comment by antonmedv** [comment]
I actually not understand what original issues is was about: https://github.com/expr-lang/expr/issues/568

Is to be able to use reuse operator names as functions:

```go
	options := []expr.Option{
		expr.Function(
			"contains",
			func(params ...any) (any, error) {
				return false, nil
			},
			new(func() bool),
		),
	}

	program, err := expr.Compile(`contains()`, options...)
```

I think this is a bad practice and not should be allowed. Also I think it's bad decision to allow to use some operators as functions =/

Let's not processed this this PR @ckganesan?

**Comment by Q16G** [comment]
operator becomes a function to increase the extensibility, convenient for users to customize or extend. Consider a scenario where the user needs to pass in multiple arguments to the contains function, and the original operator is not satisfied, we can extend it to contain from an operator to a function

**Comment by antonmedv** [comment]
> Consider a scenario where the user needs to pass in multiple arguments to the contains function

Can you show some examples?

**Comment by Q16G** [comment]
Will this operation be allowed

**Comment by antonmedv** [comment]
No. For now, not. I don't see a use case. And adding such feature complicates parser and type checker.

**Pull request closed without merging**